### PR TITLE
Make win_groupadd module behave like other modules

### PR DESCRIPTION
**Pull request by twangboy**
### What does this PR do?
Fixes issues with win_groupadd module where it was returning dictionaries instead of boolean values.
Adds logging
Returns CommandExecutionError on failure

### What issues does this PR fix or reference?
https://github.com/saltstack/salt/pull/49794

### Tests written?
Yes

### Commits signed with GPG?
Yes

### CODE REVIEW

**Comment by rallytime** [comment]
@twangboy There are some related tests failing on windows here. Can you take a look?

**Comment by damon-atkins** [comment]
Raise CommandExecutionError how it should be, but it would need all other other group modules to be updated, to behavior the same way. states/group to support it.  Is with CommandExecutionError  is you can pass back why it failed, rather than just false. Most likely raise change should be in develop.

**Comment by twangboy** [comment]
@damon-atkins Created an issue to add CommandExecutionError on failure.
https://github.com/saltstack/salt/issues/50096

**Review by dwoz — APPROVED** [review]

File: salt/modules/win_groupadd.py [hunk 1]
```diff
@@ -100,36 +103,29 @@ def add(name, **kwargs):
...
-                friendly_error = com_err.excepinfo[2].rstrip('\r\n')
-            ret['comment'] = 'Failed to create group {0}. {1}' \
-                             ''.format(name, friendly_error)
+            new_group = comp_obj.Create('group', name)
+            new_group.SetInfo()
+            log.debug('Successfully created group {0}'.format(name))
+        except pywintypes.com_error as exc:
+            msg = 'Failed to create group {0}. {1}'.format(
+                name, win32api.FormatMessage(exc.excepinfo[5]))
+            log.debug(msg)
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]: Is `debug` the right log level for an error like this?
```suggestion
            log.error(msg)
```

File: salt/modules/win_groupadd.py [hunk 2]
```diff
@@ -100,36 +103,29 @@ def add(name, **kwargs):
...
+            return False
     else:
-        ret['result'] = None
-        ret['comment'] = 'The group {0} already exists.'.format(name)
+        log.debug('The group {0} already exists.'.format(name))
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]: Same question here.

```suggestion
        log.error('The group {0} already exists.'.format(name))
```

File: salt/modules/win_groupadd.py [hunk 3]
```diff
@@ -142,38 +138,28 @@ def delete(name, **kwargs):
...
-            ret['changes'].append(('Successfully removed group {0}').format(name))
-        except pywintypes.com_error as com_err:
-            ret['result'] = False
-            if len(com_err.excepinfo) >= 2:
-                friendly_error = com_err.excepinfo[2].rstrip('\r\n')
-            ret['comment'] = (
-                    'Failed to remove group {0}.  {1}'
-                    ).format(name, friendly_error)
+            comp_obj.Delete('group', name)
+            log.debug('Successfully removed group {0}'.format(name))
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
            log.error('Successfully removed group {0}'.format(name))
```

File: salt/modules/win_groupadd.py [hunk 4]
```diff
@@ -142,38 +138,28 @@ def delete(name, **kwargs):
...
+        except pywintypes.com_error as exc:
+            msg = 'Failed to remove group {0}. {1}'.format(
+                name, win32api.FormatMessage(exc.excepinfo[5]))
+            log.debug(msg)
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
            log.error(msg)
```

File: salt/modules/win_groupadd.py [hunk 5]
```diff
@@ -142,38 +138,28 @@ def delete(name, **kwargs):
...
+            return False
     else:
-        ret['result'] = None
-        ret['comment'] = (
-                'The group {0} does not exists.'
-                ).format(name)
+        log.debug('The group {0} does not exists.'.format(name))
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
        log.warning('The group {0} does not exists.'.format(name))
```

File: salt/modules/win_groupadd.py [hunk 6]
```diff
@@ -261,54 +247,38 @@ def adduser(name, username, **kwargs):
...
-        ret['comment'] = 'Failure accessing group {0}. {1}' \
-                         ''.format(name, friendly_error)
-        return ret
-
-    existingMembers = [_get_username(x) for x in groupObj.members()]
+        group_obj = _get_group_object(name)
+    except pywintypes.com_error as exc:
+        msg = 'Failed to access group {0}. {1}'.format(
+            name, win32api.FormatMessage(exc.excepinfo[5]))
+        log.debug(msg)
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
        log.error(msg)
```

File: salt/modules/win_groupadd.py [hunk 7]
```diff
@@ -261,54 +247,38 @@ def adduser(name, username, **kwargs):
...
 
     try:
-        if username not in existingMembers:
-            if not __opts__['test']:
-                groupObj.Add('WinNT://' + username.replace('\\', '/'))
-
-            ret['changes']['Users Added'].append(username)
+        if username not in existing_members:
+            group_obj.Add('WinNT://' + username.replace('\\', '/'))
+            log.debug('Added user {0}'.format(username))
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
            log.info('Added user {0}'.format(username))
```

File: salt/modules/win_groupadd.py [hunk 8]
```diff
@@ -261,54 +247,38 @@ def adduser(name, username, **kwargs):
...
-            ret['result'] = None
-    except pywintypes.com_error as com_err:
-        if len(com_err.excepinfo) >= 2:
-            friendly_error = com_err.excepinfo[2].rstrip('\r\n')
-        ret['comment'] = (
-                'Failed to add {0} to group {1}.  {2}'
-                ).format(username, name, friendly_error)
-        ret['result'] = False
-        return ret
+            log.debug('User {0} is already a member of {1}'.format(username, name))
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
            log.info('User {0} is already a member of {1}'.format(username, name))
```

File: salt/modules/win_groupadd.py [hunk 9]
```diff
@@ -261,54 +247,38 @@ def adduser(name, username, **kwargs):
...
+    except pywintypes.com_error as exc:
+        msg = 'Failed to add {0} to group {1}. {2}'.format(
+            username, name, win32api.FormatMessage(exc.excepinfo[5]))
+        log.debug(msg)
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
        log.error(msg)
```

File: salt/modules/win_groupadd.py [hunk 10]
```diff
@@ -324,53 +294,36 @@ def deluser(name, username, **kwargs):
...
-            friendly_error = com_err.excepinfo[2].rstrip('\r\n')
-        ret['result'] = False
-        ret['comment'] = 'Failure accessing group {0}. {1}' \
-                         ''.format(name, friendly_error)
-        return ret
+        group_obj = _get_group_object(name)
+    except pywintypes.com_error as exc:
+        msg = 'Failed to access group {0}. {1}'.format(
+            name, win32api.FormatMessage(exc.excepinfo[5]))
+        log.debug(msg)
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
        log.error(msg)
```

File: salt/modules/win_groupadd.py [hunk 11]
```diff
@@ -324,53 +294,36 @@ def deluser(name, username, **kwargs):
...
 
     try:
-        if salt.utils.win_functions.get_sam_name(username) in existingMembers:
-            if not __opts__['test']:
-                groupObj.Remove('WinNT://' + username.replace('\\', '/'))
-
-            ret['changes']['Users Removed'].append(username)
+        if salt.utils.win_functions.get_sam_name(username) in existing_members:
+            group_obj.Remove('WinNT://' + username.replace('\\', '/'))
+            log.debug('Removed user {0}'.format(username))
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
            log.info('Removed user {0}'.format(username))
```

File: salt/modules/win_groupadd.py [hunk 12]
```diff
@@ -324,53 +294,36 @@ def deluser(name, username, **kwargs):
...
-        ret['comment'] = (
-                'Failed to remove {0} from group {1}.  {2}'
-                ).format(username, name, friendly_error)
-        ret['result'] = False
-        return ret
+            log.debug('User {0} is not a member of {1}'.format(username, name))
+    except pywintypes.com_error as exc:
+        msg = 'Failed to remove {0} from group {1}. {2}'.format(
+            username, name, win32api.FormatMessage(exc.excepinfo[5]))
+        log.debug(msg)
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
        log.error(msg)
```

File: salt/modules/win_groupadd.py [hunk 13]
```diff
@@ -387,78 +340,62 @@ def members(name, members_list, **kwargs):
...
-           'result': True,
-           'changes': {'Users Added': [], 'Users Removed': []},
-           'comment': []}
-
     members_list = [salt.utils.win_functions.get_sam_name(m) for m in members_list.split(",")]
     if not isinstance(members_list, list):
-        ret['result'] = False
-        ret['comment'].append('Members is not a list object')
-        return ret
+        log.debug('member_list is not a list')
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
        log.error('member_list is not a list')
```

File: salt/modules/win_groupadd.py [hunk 14]
```diff
@@ -387,78 +340,62 @@ def members(name, members_list, **kwargs):
...
-                ).format(name, friendly_error))
-        return ret
-    existingMembers = [_get_username(x) for x in groupObj.members()]
-    existingMembers.sort()
+        obj_group = _get_group_object(name)
+    except pywintypes.com_error as exc:
+        # Group probably doesn't exist, but we'll log the error
+        msg = 'Failed to access group {0}. {1}'.format(
+            name, win32api.FormatMessage(exc.excepinfo[5]))
+        log.debug(msg)
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
        log.error(msg)
```

File: salt/modules/win_groupadd.py [hunk 15]
```diff
@@ -387,78 +340,62 @@ def members(name, members_list, **kwargs):
...
-            except pywintypes.com_error as com_err:
-                if len(com_err.excepinfo) >= 2:
-                    friendly_error = com_err.excepinfo[2].rstrip('\r\n')
-                ret['result'] = False
-                ret['comment'].append((
-                        'Failed to add {0} to {1}.  {2}'
-                        ).format(member, name, friendly_error))
-                #return ret
+                obj_group.Add('WinNT://' + member.replace('\\', '/'))
+                log.debug('User added: {0}'.format(member))
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
                log.info('User added: {0}'.format(member))
```

File: salt/modules/win_groupadd.py [hunk 16]
```diff
@@ -387,78 +340,62 @@ def members(name, members_list, **kwargs):
...
+            except pywintypes.com_error as exc:
+                msg = 'Failed to add {0} to {1}. {2}'.format(
+                    member, name, win32api.FormatMessage(exc.excepinfo[5]))
+                log.debug(msg)
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
                log.error(msg)
```

File: salt/modules/win_groupadd.py [hunk 17]
```diff
@@ -387,78 +340,62 @@ def members(name, members_list, **kwargs):
...
-                    friendly_error = com_err.excepinfo[2].rstrip('\r\n')
-                ret['result'] = False
-                ret['comment'].append((
-                        'Failed to remove {0} from {1}.  {2}'
-                        ).format(member, name, friendly_error))
-                #return ret
-
-    return ret
+                obj_group.Remove('WinNT://' + member.replace('\\', '/'))
+                log.debug('User removed: {0}'.format(member))
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
                log.info('User removed: {0}'.format(member))
```

File: salt/modules/win_groupadd.py [hunk 18]
```diff
@@ -387,78 +340,62 @@ def members(name, members_list, **kwargs):
...
+            except pywintypes.com_error as exc:
+                msg = 'Failed to remove {0} from {1}. {2}'.format(
+                    member, name, win32api.FormatMessage(exc.excepinfo[5]))
+                log.debug(msg)
```
Comment (cachedout) [inline salt/modules/win_groupadd.py]:
```suggestion
                log.error(msg)
```

**Review by cachedout — CHANGES_REQUESTED** [review]

**Comment by rallytime** [comment]
@twangboy Did you see the feedback above?

**Comment by twangboy** [comment]
@rallytime Yes, I added it to my list

**Comment by twangboy** [comment]
@cachedout Added your changes and rebased

**Comment by twangboy** [comment]
Updated the branch... Is this OK now?

**Comment by rallytime** [comment]
@cachedout Can you review this one more time?

**Comment by twangboy** [comment]
@cachedout @rallytime Rebased...

**Review by cachedout — APPROVED** [review]

**Pull request merged**